### PR TITLE
Adding CICD task to generate rustdocs both manually or as pipeline job

### PR DESCRIPTION
**Pull request by jimmychu0807**
This PR added two jobs `rebuild-all-rustdoc` and `republish-all-rustdoc` to manually trigger a (re-)publishing rustdocs of all versions specified in `RUSTDOCS_DEPLOY_REFS`. 

When manually triggered, additional variable `REPUBLISH_ALL_RUSTDOCS` need to be set to `true`. [This pipeline](https://gitlab.parity.io/parity/substrate/-/pipelines/159127) shows that the two new jobs run successfully (I have the running condition set on PR # for testing).

This should be run at least once by ops staff after a merge to `master`.

Currently only rustdocs of `master` and `monthly-2021-09+1` are deployed because running `cargo +nightly doc` with `monthly-2021-08` and prior reports error on `prost-derive` crate.

@TriplEight @rcny Could you have a look at this PR? Thanks

### CODE REVIEW

**Comment by jimmychu0807** [comment]
@TriplEight I have gotten the latest `master` change and merge to this branch, and checked the rustdocs publication process works ([see this workflow](https://gitlab.parity.io/parity/substrate/-/pipelines/164942), particularly on the [`build-rustdoc`](https://gitlab.parity.io/parity/substrate/-/jobs/1208672) and [`publish-rustdoc`](https://gitlab.parity.io/parity/substrate/-/jobs/1208677) jobs).

If you have time this week, please review and see if the PR is good to merge. So I can then publish the Substrate rustdocs for `monthly-2021-10` and `monthly-2021-11-1` tags. Thanks.

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -159,6 +169,19 @@ default:
     # this job runs only on nightly pipeline with the mentioned variable, against `master` branch
     - if: $CI_COMMIT_REF_NAME == "master" && $CI_PIPELINE_SOURCE == "schedule" && $PIPELINE == "nightly"
 
+.build-rustdoc-rules:              &build-rustdoc-rules
+  rules:
+    - if: $CI_PIPELINE_SOURCE == "pipeline"
+      when: never
+    - if: $CI_PIPELINE_SOURCE == "web" && $RM_RUSTDOC_REF
```
Comment (TriplEight) [inline .gitlab-ci.yml:176]:
```suggestion
    - if: $CI_PIPELINE_SOURCE == "schedule" && $RM_RUSTDOC_REF
```

It's better to create some preset schedules https://gitlab.parity.io/parity/substrate/-/pipeline_schedules with variables and branches and set them to manual (not active). And explain it in the attached doc. Also it's possible to run just rusdocs from those "manual schedules" jobs without running all the other CI.

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -159,6 +169,19 @@ default:
...
+      when: never
+    - if: $CI_PIPELINE_SOURCE == "schedule" || $CI_PIPELINE_SOURCE == "push" ||
+        ($CI_PIPELINE_SOURCE == "web" && $BUILD_RUSTDOC_REF)
```
Comment (TriplEight) [inline .gitlab-ci.yml:179]:
```suggestion
    - if: $CI_PIPELINE_SOURCE == "schedule" && $BUILD_RUSTDOC_REF
```
explained above, plus `$CI_PIPELINE_SOURCE == "push"` does not work in the mirrored setup.

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -159,6 +169,19 @@ default:
...
+
+.rm-rustdoc-rules:                 &rm-rustdoc-rules
+  rules:
+    - if: $CI_PIPELINE_SOURCE == "web" && $RM_RUSTDOC_REF
```
Comment (TriplEight) [inline .gitlab-ci.yml:183]:
```suggestion
    - if: $CI_PIPELINE_SOURCE == "schedule" && $RM_RUSTDOC_REF
```
explained above

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -781,79 +854,54 @@ publish-rustdoc:
...
-    # publish though, hence the `|| true`.
-    - git commit -m "___Updated docs for ${CI_COMMIT_REF_NAME}___" ||
-        echo "___Nothing to commit___"
-    - git push origin gh-pages --force
-  after_script:
-    - rm -rf .git/ ./*
+    - '[[ -e "$LATEST" ]] && { [[ -h latest ]] && rm latest; ln -sf "${LATEST}" latest; }'
+    - *git-force-push-rm-script
+
+rm-rustdoc:
```
Comment (TriplEight) [inline .gitlab-ci.yml:884]: it's an unnecessary complication to hold this job in the CI config, it's not automation when you have a job that's going to be triggered rarely manually.
It's better to describe the process in the doc. Or put the whole rustdoc update logic into a separate script and condition its' use with input variables. Say, check first for `$RM_RUSTDOC_REF` then remove and exit.

File: docs/rustdoc-release.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
+# Substrate Rustdoc Release Process
+
+Substrate rustdoc sits on the [paritytech/substrate `gh-pages` branch](https://github.com/paritytech/substrate/tree/gh-pages) and is accessible at https://paritytech.github.io/substrate.
```
Comment (TriplEight) [inline docs/rustdoc-release.md:3]: Was about to ask what sources rustdocs on substrate.dev. But hey, it's great that now it [points](https://docs.substrate.io/rustdocs/) here!

File: docs/rustdoc-release.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+If you inspect the `gh-pages` branch, it is just various different version of rustdocs located in different folders, with folder name marking the monthly tag version, or a release tag version.
```
Comment (TriplEight) [inline docs/rustdoc-release.md:5]:
```suggestion
If you inspect the `gh-pages` branch, it is just various different versions of rustdocs located in different folders, with folder names marking the monthly tag version, or a release tag version.
```

Comment (TriplEight) [reply docs/rustdoc-release.md:5]: also not sure if we check the line length for the mds, but we should.

File: docs/rustdoc-release.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+The actual rustdoc generation process takes place as part of the project [GitLab pipeline](https://gitlab.parity.io/parity/substrate/-/pipelines).
```
Comment (TriplEight) [inline docs/rustdoc-release.md:7]:
```suggestion
The actual rustdoc generation process takes place as part of the projects' [GitLab pipeline](https://gitlab.parity.io/parity/substrate/-/pipelines).
```

File: .gitlab-ci.yml [hunk 8]
```diff
@@ -192,6 +215,52 @@ default:
...
+.git-force-push-rm-script:         &git-force-push-rm-script
+  # push it back to github gh-pages
+  - git add --all --force
+  - git commit -m "___Updated docs for ${CI_COMMIT_REF_NAME}___" || echo "___Nothing to commit___"
+  - git push origin gh-pages --force
+  - rm -rf .git/ ./*
+
+.test-build-rustdoc-script:        &test-build-rustdoc-script
+  # The actual check on whether we need to build rustdocs. This task will run if it is one of
+  #   branch/tag in CONT_RUSTDOC_BUILT_REFS or triggered via web with BUILD_RUSTDOC_REF set to
```
Comment (TriplEight) [inline .gitlab-ci.yml:256]: Also relates to https://github.com/paritytech/substrate/pull/9856/files#diff-037ea159eb0a7cb0ac23b851e66bee30fb838ee8d0d99fa331a1ba65283d37f7R244

We are mirroring our repos the [following way](https://github.com/paritytech/devops/wiki/Gitlab:-Mirroring--setup,-CICD-Pipeline-for-Github-repos#important):
```
        fetch = +refs/heads/master:refs/remotes/upstream/master
        fetch = +refs/pull/*/head:refs/remotes/upstream/*
        fetch = +refs/tags/*:refs/tags/*
```

It means that only `master` branch keeps its' name, all the PRs are mirrored by their #. I.e. this one is #9856 so you can find your pipelines in https://gitlab.parity.io/parity/substrate/-/pipelines/164942 and code in https://gitlab.parity.io/parity/substrate/-/tree/9856.
This means that in `CONT_RUSTDOC_BUILT_REFS` you can mention either `master`, or tags, or PR numbers.

It works with the default `master` value, but how will this work with the other value?

File: docs/rustdoc-release.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+## Add a new version of rustdoc
+
+To add a new version of rustdocs, you need to manually trigger a build job in the
+[GitLab pipeline](https://gitlab.parity.io/parity/substrate/-/pipelines/new).
```
Comment (TriplEight) [inline docs/rustdoc-release.md:12]:
```suggestion
To add a new version of rustdocs, you need to manually trigger a schedule pipeline in the
[GitLab schedules](https://gitlab.parity.io/parity/substrate/-/pipeline_schedules).
```

File: docs/rustdoc-release.md [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+## Remove an existing version of rustdoc
+
+To remove an existing version of rustdocs, you need to manually trigger a build job in the
+[GitLab pipeline](https://gitlab.parity.io/parity/substrate/-/pipelines/new).
+
+In the pipeline screen, pick:
+
+- Run for branch name or tag: `master`
+
+- Add the following variables:
```
Comment (TriplEight) [inline docs/rustdoc-release.md:49]:
```suggestion
- Add the following variable:
```

File: docs/rustdoc-release.md [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+  - `RM_RUSTDOC_REF`: `<git branch name | tag name>` (required)
+    This parameter specifies the version of rustdoc to be removed.
+
+- Wait for the whole pipeline to complete. This take 10 - 20 mins.
+
+## Config branch/tag for rebuilding rustdoc when there is new commit
+
+In the [`.gitlab-ci.yml`](../.gitlab-ci.yml), there is [`CONT_RUSTDOC_BUILT_REFS` variable](../.gitlab-ci.yml#L52). Tags branches specified in this var will be rebuilt everytime when new
```
Comment (TriplEight) [inline docs/rustdoc-release.md:57]:
```suggestion
In the [`.gitlab-ci.yml`](../.gitlab-ci.yml), there is [`CONT_RUSTDOC_BUILT_REFS` variable](../.gitlab-ci.yml#L52).
Tags and branches specified in this var will be rebuilt every time when new
```

File: docs/rustdoc-release.md [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+commits are pushed. Currently it is set to `master` only. It supports multiple value by separating
```
Comment (TriplEight) [inline docs/rustdoc-release.md:58]:
```suggestion
commits are pushed. Currently, it is set to `master` only. It supports multiple values by separating
```

**Review by TriplEight — CHANGES_REQUESTED** [review]

**Comment by bkchr** [comment]
@jimmychu0807 for what do we need this?

**Comment by jimmychu0807** [comment]
> @jimmychu0807 for what do we need this?

@bkchr This PR makes:

1. Make the repo rustdocs support multiple versions. For [Substrate](https://paritytech.github.io/substrate/) there are multiple versions of rustdoc deployed and that specific index page linked to, is also built as part of this CI workflow.

File: .gitlab-ci.yml (hunk 8 above)
Comment (jimmychu0807) [reply .gitlab-ci.yml:254]: When a new pull request arrive, I don't foresee it to trigger this. It is only when a PR is merging to a certain branch (`master`), or a certain branch or tag name is pushed that we will trigger the rustdoc build.

**Comment by NukeManDan** [comment]
Guys: I think we ***should not*** host anything but master here on the repo. Why? It's BIG in the git size to host them!! Instead we should host all versioned archives in https://github.com/substrate-developer-hub/rustdocs and only maintain the active master here.

**Comment by NukeManDan** [comment]
https://github.com/paritytech/substrate/tree/gh-pages IMHO should be copied locally, and some git surgery done to completely prune it from the history before starting it once again with _only_ master `/target/doc` copied over via CI, _nothing else_. 

We can use https://docs.substrate.io/rustdocs/ to link to pages appropriately.

**Review by NukeManDan — CHANGES_REQUESTED** [review]
https://github.com/paritytech/substrate/tree/gh-pages IMHO should be copied locally, and some git surgery done to completely prune it from the history before starting it once again with _only_ master `/target/doc` copied over via CI, _nothing else_. 

We can use https://docs.substrate.io/rustdocs/ to link to pages appropriately.

**Comment by TriplEight** [comment]
@NukeManDan to be fair

```bash
$ git rev-list --disk-usage --objects HEAD..gh-pages
277437609
```

Is **not that** much, but only compared to 7.5 GB.
I assume this PR was done in assumption to store the state of the documentation and host it straight away from the `gh-pages` branch. But for a fatter repo we might re-thing this approach.

**Comment by jimmychu0807** [comment]
I think it is better to host the rustdoc in Substrate repo. It should be less maintenance in the long-run, and it is good the rustdoc sits closing to the src code.

**Comment by jimmychu0807** [comment]
let me think about this, bot.

**Pull request closed without merging**